FUN-1355 adding ability for users to map fields dynamically

Alexalex1432 opened the pull request.

Commit:
FUN-1355 adding ability for users to map fields dynamically from integration

Alexalex1432 commented:
The managed package link if you'd like to install this while it's under review: https://login.salesforce.com/packaging/installPackage.apexp?p0=04t15000000OBNf

Commit:
FUN-1355 Adding additional properties for dedication and operationsTipAmount

adelorme82 commented:
I have not tested or reviewed this in depth, am waiting for more detailed instructions on how to migrate from unmanaged package to managed package

Commit:
FUN-1355 adding ability to provide constant values for mappings instead of dynamic values from funraise.  Useful for things like StageName/Probability on Opportunity

Alexalex1432 commented:
@adelorme82 Okay, per our discussion elsewhere, the migration steps are straight forward with no additional manual work needed.  The namespacing takes care of most of the concerns with conflicts, and the data migration will just be dataloader from old fields to new fields for customers.  Here is the link for the latest beta version of the package: https://login.salesforce.com/packaging/installPackage.apexp?p0=04t15000000OBPR

adelorme82 commented on src/classes/frModel.cls:
```diff
@@ -0,0 +1,71 @@
+        getObject().put('fr_ID__c', String.valueOf(request.get('id')));
+    }
+
+    private static void write(SObject record, Map<String, Schema.SObjectField> fields, String fieldName, Object value) {
```
@Alexalex1432 what do you think about adding some `handleExceptions()` here to be more verbose about when field writes fail?

Alexalex1432 replied:
All of the calls inside here (with the exception of the id case on line 25, which isn't used currently since salesforce ids aren't in funraise) call the other write() method that has the try catch with the error record insert, I think it covers all the possible exceptions.  Where else were you thinking of adding some handleExceptions and what additional info should we catch?

adelorme82 commented on src/classes/frModel.cls:
```diff
@@ -0,0 +1,71 @@
+
+        Map<String, frMapping__c> frNameToSfName = new Map<String, frMapping__c>();
+        for(frMapping__c mapping : getMappings()) {
+            frNameToSfName.put(mapping.fr_Name__c, mapping);
```
Are you opposed to logging some message here if the mapping.fr_Name__c is empty? Something went wrong when I was setting up my package and all of my mappings had no funraise fields, but it worked on the setup page and SFDC did not complain or generate any logs when I sent over a donation. It also didn't actually create the donation though so there is at least some indication that things went wrong

Alexalex1432 replied:
Yeah this is a good catch.  I'll take another look at this entire method because this part especially doesn't jive well with constant mappings, since those ones dont need a funraise field

adelorme82 replied:
Actually turns out that it did create the opportunity, but it exposed something a bit weird. The fr_id__c field exists from the old version and in the new version, and the queries to find donors by fr id are all looking at the new version. Is this something we want to do anything about or is this just a "gotta data loader it, fool" type issue to move the funraise ids from the old column to the new one?

Alexalex1432 replied:
The old version of the field should be fr_Funraise_Id__c

Alexalex1432 replied:
As far as looking at the different fields, I'm not sure we can do much about it in code (like if we try to query for the old version of the field but that field doesn't even exist I think we'd get some SF exceptions), so yeah I think it is a you gotta data loader the data

adelorme82 commented on src/classes/frWSDonationController.cls:
```diff
@@ -38,29 +38,30 @@
 
         resp.put('id',donation.getOpportunityId());
         resp.put('errors',errors);
+        resp.put('result',result);
```
Thoughts on renaming this to `success` instead?

Alexalex1432 replied:
Definitely makes sense, will make that change tomorrow along with any other feedback gathered

Commit:
FUN-1355 addressing code review feedback.  

Fixing frModel building the mapping map (heh) based on funraise field
name since constants do not have funraise field values in the mapping
record

Alexalex1432 commented:
Here's the link to the newest beta package (note you have to uninstall the previous version of the beta package before installing this one): https://login.salesforce.com/packaging/installPackage.apexp?p0=04t15000000OBQ5  The changes in this package is the feedback pushed in the latest commit (	c485a0b) and also including the custom metadata records which I did not know you had to explicitly do
